flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java

```diff
@@ -0,0 +1,517 @@
+		final TestingDispatcherGateway testingDispatcherGateway = new TestingDispatcherGateway.Builder()
+			.setSubmitFunction(
+				submittedJob -> {
+					submittedJobFuture.complete(submittedJob);
+					return CompletableFuture.completedFuture(Acknowledge.get());
+				})
+			.build();
+
+		dispatcherServiceFactory = createDispatcherServiceFactoryFor(testingDispatcherGateway);
+
+		try (final SessionDispatcherLeaderProcess dispatcherLeaderProcess = createDispatcherLeaderProcess()) {
+			dispatcherLeaderProcess.start();
+
+			// wait first for the dispatcher service to be created
+			dispatcherLeaderProcess.getDispatcherGateway().get();
+
+			jobGraphStore.putJobGraph(JOB_GRAPH);
+			dispatcherLeaderProcess.onAddedJobGraph(JOB_GRAPH.getJobID());
+
+			final JobGraph submittedJobGraph = submittedJobFuture.get();
+
+			assertThat(submittedJobGraph.getJobID(), is(JOB_GRAPH.getJobID()));
+		}
+	}
+
+	@Test
+	public void onAddedJobGraph_ifNotRunning_isBeingIgnored() throws Exception {
+		final CompletableFuture<JobID> recoveredJobFuture = new CompletableFuture<>();
+		jobGraphStore = TestingJobGraphStore.newBuilder()
+			.setRecoverJobGraphFunction(
+				(jobId, jobGraphs) -> {
+					recoveredJobFuture.complete(jobId);
+					return jobGraphs.get(jobId);
+				})
+			.build();
+
+		try (final SessionDispatcherLeaderProcess dispatcherLeaderProcess = createDispatcherLeaderProcess()) {
+			dispatcherLeaderProcess.start();
+
+			// wait until the process has started the dispatcher
+			dispatcherLeaderProcess.getDispatcherGateway().get();
+
+			// now add the job graph
+			jobGraphStore.putJobGraph(JOB_GRAPH);
+
+			dispatcherLeaderProcess.closeAsync();
+
+			dispatcherLeaderProcess.onAddedJobGraph(JOB_GRAPH.getJobID());
+
+			try {
+				recoveredJobFuture.get(10L, TimeUnit.MILLISECONDS);
+				fail("onAddedJobGraph should be ignored if the leader process is not running.");
+			} catch (TimeoutException expected) {}
+		}
+	}
+
+	@Test
+	public void onAddedJobGraph_failingRecovery_propagatesTheFailure() throws Exception {
+		final FlinkException expectedFailure = new FlinkException("Expected failure");
+		jobGraphStore = TestingJobGraphStore.newBuilder()
+			.setRecoverJobGraphFunction(
+				(ignoredA, ignoredB) -> {
+					throw expectedFailure;
+				})
+			.build();
+
+		try (final SessionDispatcherLeaderProcess dispatcherLeaderProcess = createDispatcherLeaderProcess()) {
+			dispatcherLeaderProcess.start();
+
+			// wait first for the dispatcher service to be created
+			dispatcherLeaderProcess.getDispatcherGateway().get();
+
+			jobGraphStore.putJobGraph(JOB_GRAPH);
+			dispatcherLeaderProcess.onAddedJobGraph(JOB_GRAPH.getJobID());
+
+			final CompletableFuture<Throwable> errorFuture = fatalErrorHandler.getErrorFuture();
+			final Throwable throwable = errorFuture.get();
+			Assert.assertThat(ExceptionUtils.findThrowable(throwable, expectedFailure::equals).isPresent(), Is.is(true));
```

Comment:
True, it should be consistent. I will update the test case.

Why do you prefer JUnit's `assertThat` method?

Reply:
The only reason is to keep consistency so that all assertions come from `org.junit.Assert`, e.g., `assertTrue`, `assertEquals`, `assertThat`). Semantically there is no difference at the moment.